### DISCUSSION TITLE
ThirdPartyExternalLink feature requests (update children prop, add target prop)

### DISCUSSION THREAD

**Discussion by billiedoescode** [CMSgov/design-system#3009] [Ideas]
Is it possible to update the [ThirdPartyExternalLink](https://design.cms.gov/storybook/?path=/docs/components-thirdpartyexternallink--docs) to...

1. Accept ReactNodes/other components as valid input to the `children` prop?
We're currently forced to use `// @ts-expect-error: TS2322` because we have third-party external links using things like a social media fa-icon, an image, buttons, etc, and we need more customization in how we display those links than just a `string`. The component in its current state will _still_ work, so is it possible to remove the `string` qualifier?
1. Not to start an argument over whether having links automatically open in a new tab is correct or not...but it would be helpful to allow the user to pass a `target` prop regardless.

Thank you! I wasn't sure if these choices were intentional design decisions or not.

**Comment by pwolfert** [comment]
Hi @billie-rose, let's unpack your use cases some more.

> we have third-party external links using things like a social media fa-icon, an image, buttons, etc

this sounds like you want to be able to customize more than just the content inside the link but the anchor element itself. Is that right?

**Reply by billiedoescode** [reply]
Not exactly (I think?), though not opposed to that. I'll point to the social media icons we have in the healthcare.gov footer as an example:

![ex-modal](https://github.com/CMSgov/design-system/assets/5013529/cf2f57ad-089b-4ed8-b162-98e99d854e62)

These icons are external links and should prompt the external link modal, but the links are obviously not just text strings. We had to add some simple logic/css to hide the default icon (
![exlicon](https://github.com/CMSgov/design-system/assets/5013529/989c33c7-4b8b-49ad-a0b5-a62058080d83)) in these cases. Not ideal, but fine. I understand that this component isn't meant to be used this way but we wanted to keep the modal experience as consistent as possible.

The main issue though is that our TS compiler is unhappy with us passing non-string children to the component's `children` prop (E.g., we're using our social media icons & their relevant elements as children instead of plain text). Is it possible to be more lenient here with this property's typing, given that it still loads properly given children of different types?

**Comment by pwolfert** [comment]
> Not to start an argument over whether having links automatically open in a new tab is correct or not...

It's possible the only thing we'd want to add here if opening a new tab is letting the user know in the dialog text that it's going to open a new tab. I can run that by an accessibility specialist.

**Comment by krithimurthy** [comment]
Hi @billie-rose and @seesleestak! Thanks for bringing this to our team's attention. After speaking with our CMS PM, we're going to hold off on making customizations for the [ThirdPartyExternalLink](https://design.cms.gov/storybook/?path=/docs/components-thirdpartyexternallink--docs), as our current implementation follows current CMS guidance and there haven't been multiple use cases identified to officially adopt this change in the design system just yet.

For the time being, we recommend your team make the dialog locally in your application (tagging @pwolfert to provide any other technical details, if necessary). We also suggest engaging with the UX leads (Allyssa & Erik) for further discussion if the current implementation is impacting user experience. 

Thank you!

**Reply by seesleestak** [reply]
Hey @krithimurthy, we'd recently done work to migrate _away_ from our custom version of the third party external link modal which I think is what prompted this feedback. I'd assumed since the component was in the DS that our team should be consolidating around its usage rather than our version. Is that not the case?

**Comment by DavidMCarek** [comment]
@krithimurthy Just wanted to ask about this again since it sounds like there was some discussion that happened after this. The only change required would be to update the `type` of the children property from `string` to `string | React.Node`. Currently it is possible to use the `ThirdPartyExternalLink` component in this way by telling typescript to ignore the type check on children. So, it seems unnecessarily restrictive to limit  the `children` property to just a string. If there is a case to be made for only using a string for the external link then at the very least we should expose the dialog as a separate component.

**Comment by sydk-adhoc** [comment]
👋 from MCT. I just found this discussion, and also found the component for the first time. One of my teammates was creating an external link simply by following the pattern in the component code for the link that opens the dialog (you just wrap an `a` or React `Link` around an `ExternalLinkIcon`). So we don't open a dialog at all. We also have `target="_blank"` and `rel="noreferrer"` on our `a` tag.

It seems there wasn't any follow-up on this, but wondering if any updates would be considered. In our case, it would be standard `HTMLAnchorElement `attributes either being added by default to the "Continue" `Button` (with `href`, renders as `a`, takes all attributes), or passing through any allowable attributes to that component.

Thanks
